challenge-10/challenge-10.js

```diff
@@ -1,113 +1,204 @@
-divisão e resto. Crie variáveis com os nomes "subtraction",
-"multiplication", "division" e "mod".
-*/
-// ?
-
-/*
-Repita o PASSO 2 novamente, mas passando um operador inválido, para ver se
-a mensagem de erro será mostrada no console.
-*/
-// ?
\ No newline at end of file
+(function () {
+
+  /*
+  Crie uma IIFE que envolva todo esse arquivo (inclusive esse comentário),
+  e faça a indentação correta.
+  */
+
+  /*
+  Sem alterar os códigos nos `console.log` abaixo, faça com que o retorno
+  deles seja "true", usando os Wrapper Objects como "conversores" nos valores
+  das variáveis. Analise o que está sendo impresso no console para saber como
+  resolver o problema corretamente.
+  */
+  var five = Number('5');
+  console.log( five + ' é número?', typeof five === 'number' );
+
+  var concat = String(10) + String(10);
+  console.log( '"' + concat + '" é uma string? E é igual a "1010"?', typeof concat === 'string' && concat === '1010' );
+
+  /*
+  Voltando ao exemplo da calculadora, vamos utilizar mais uma abordagem
+  funcional, mas dessa vez, separando algumas responsabilidades.
+  - Primeiro, crie um objeto chamado `operation` que terá as propriedades:
+  '+', '-', '*', '/' e '%'.
+  - Cada propriedade vai receber uma função (logo, elas serão métodos), e essa
+  função receberá dois parâmetros e retornará a operação referente à sua
+  propriedade, usando os valores passados por parâmetro.
+  */
+  var operation = {
+    '+': function (n1, n2) { return n1 + n2; },
+    '-': function (n1, n2) { return n1 - n2; },
+    '*': function (n1, n2) { return n1 * n2; },
+    '/': function (n1, n2) { return n1 / n2; },
+    '%': function (n1, n2) { return n1 % n2; }    
+  };
+
+  /*
+  Crie uma função chamada `isOperatorValid`, que receberá um operador por
+  parâmetro.
+  - Essa função será responsável por verificar se o operador passado por
+  parâmetro a ela é válido, ou seja, se ele é igual a '+', '-', '*', '/' ou
+  '%'.
+  - Se for igual a qualquer um desses, ela deverá retornar "true".
+  Caso contrário, "false".
+  - O desafio é fazer o retorno sem usar "if" ou "switch".
+  */
+  
+   function isOperatorValid(operator) {
+    return operation[operator] !== undefined;
+  };
+  
+  /*
+  Agora vamos criar a calculadora.
+  - Crie uma função chamada `calculator`, que receberá como parâmetro um
+  operador;
+  - Se o operador não for válido, a função deve retornar "false";
+  - Se o operador for válido, retornar uma segunda função que receberá dois
+  parâmetros;
+  - Se algum dos parâmetros não for um número, retornar "false";
+  - Senão, retornar o método do objeto "operation" criado acima, baseado no
+  operador passado para a função "calculator", e passando para esse método
+  os dois parâmetros da função de retorno de "calculator".
+  */
+  function calculator(operator) {
+    if (!isOperatorValid(operator)) {
+      return false
+    }
+    return function (n1, n2) {
+      if (typeof n1 !== 'number' && typeof n2 !== 'number') {
```

Comment:
-1
Do jeito que você fez, para que essa expressão toda seja verdadeira (ou seja, pra entrar no if), os dois argumentos precisam ser diferentes de número.

Para entender melhor, vamos fazer alguns testes.

- A primeira coisa que precisamos saber é que, quando usamos `&&`, para que a expressão toda retorne `true`, todas as expressões testadas precisam retornar `true`. Se ao menos uma delas for `false`, o retorno da expressão será `false`.

- Ou seja, se **resultado do if** for `true`, o interpretador vai entrar no `if` e retornar `false`. É o que queremos se **qualquer um** dos argumentos for diferente de número. Caso os dois argumentos sejam número, o **resultado do if** será `false`, e o código continuará normalmente.

Analise a tabela abaixo:

n1  | n2   | typeof n1 !== 'number' | typeof n1 !== 'number'   | resultado do if
:--: | :----: | ----------------------------- | ------------------------------- | ----------------------
1   | 2      | false                            | false                               | false (não entra no `if`)
'a' | 1      | true                              | false                                | false (não entra no `if`)
'a' | 'b'     | true                             | true                                 | true (entra no `if`)

Olhe o segundo caso: eu entrei com uma `string` e com um `number`, e o **resultado do if** foi `false`! Logo, não vai entrar no `if`, e vai acabar executando a operação, resultando em um resultado diferente do esperado.

Para que a nossa calculadora funcione corretamente, precisamos testar se **qualquer um** dos valores entrados é diferente de número, e retornar `false`, caso isso seja verdadeiro.

Para resumir: 
- se precisamos testar **qualquer um** dos valores, independente qual, usamos `||` (OR);
- se precisamos testar **todos os valores**, usamos `&&` (AND).

Ficou tranquilo de compreender? Se tiver dúvidas quanto a essa parte me avise :)